lte/cloud/go/services/had/servicers/servicer.go

```diff
@@ -0,0 +1,263 @@
+	"magma/orc8r/cloud/go/services/state"
+	"magma/orc8r/cloud/go/services/state/wrappers"
+	"magma/orc8r/lib/go/metrics"
+	orc8rprotos "magma/orc8r/lib/go/protos"
+
+	"github.com/golang/glog"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	validSecsSinceStateReported = 180
+)
+
+type HAServicer struct {
+	promClient query_api.PrometheusAPI
+}
+
+// NewHAServicer creates a new service implementing the HaD proto file
+func NewHAServicer(promClient query_api.PrometheusAPI) protos.HaDServer {
+	return &HAServicer{
+		promClient: promClient,
+	}
+}
+
+// GetEnodebOffloadState fetches all primary gateways that the calling gateway
+// is in a gateway pool with. For each of these gateways, it then fetches the
+// offload state for each of the gateway's ENBs.
+func (s *HAServicer) GetEnodebOffloadState(ctx context.Context, request *protos.GetEnodebOffloadStateRequest) (*protos.GetEnodebOffloadStateResponse, error) {
+	ret := &protos.GetEnodebOffloadStateResponse{}
+	secondaryGw := orc8rprotos.GetClientGateway(ctx)
+	if secondaryGw == nil {
+		return ret, status.Errorf(codes.PermissionDenied, "Missing Gateway Identity")
+	}
+	if !secondaryGw.Registered() {
+		return ret, status.Errorf(codes.PermissionDenied, "Gateway is not registered")
+	}
+	cfg, err := configurator.LoadEntityConfig(secondaryGw.GetNetworkId(), lte.CellularGatewayEntityType, secondaryGw.LogicalId, lte_models.EntitySerdes)
+	if err != nil {
+		return ret, err
+	}
+	cellularCfg, ok := cfg.(*lte_models.GatewayCellularConfigs)
+	if !ok {
+		return ret, status.Errorf(codes.Internal, "could not convert stored config to type GatewayCellularConfigs for gw %s", secondaryGw.LogicalId)
+	}
+	if cellularCfg.Pooling == nil || len(cellularCfg.Pooling) == 0 {
+		return ret, fmt.Errorf("Gateway '%s' is not configured in a gateway pool", secondaryGw.LogicalId)
+	}
+
+	// All gateway pool records must have the same capacity, so use the first
+	// entry
+	callingRelativeCapacity := cellularCfg.Pooling[0].MmeRelativeCapacity
+	gwIDsToEnbs := map[string][]string{}
+	for _, record := range cellularCfg.Pooling {
+		gwIDsToEnbsInPool, err := s.getPrimaryGatewaysToEnodebs(secondaryGw.GetNetworkId(), string(record.GatewayPoolID), callingRelativeCapacity)
+		if err != nil {
+			return &protos.GetEnodebOffloadStateResponse{}, err
+		}
+		// Since a gateway can be in multiple pools, it is possible
+		// there could be key collisions. Since the ENBs values will be the
+		// same each time, this is okay.
+		for k, v := range gwIDsToEnbsInPool {
+			gwIDsToEnbs[k] = v
+		}
+	}
+	glog.V(2).Infof("Found the following primary gatewayIDs to ENB SNs: %v", gwIDsToEnbs)
+
+	enbSNsToOffloadState := map[uint32]protos.GetEnodebOffloadStateResponse_EnodebOffloadState{}
+	for primaryGwID, enbs := range gwIDsToEnbs {
+		isCheckinValid, err := s.isGatewayCheckinValid(secondaryGw.NetworkId, primaryGwID)
+		if err != nil {
+			glog.Error(err)
```

Comment:
Here and generally: when we log an error instead of propagating it up the call stack, add a comment describing the decision